### PR TITLE
Subscriptions handling improvements

### PR DESCRIPTION
**Pull request by GediminasZukas**
Fixes for: #3, #4, #5

### CODE REVIEW

File: library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt [hunk 1]
```diff
@@ -21,8 +21,15 @@ class Subscriptions constructor(private val consumer: Consumer) {
      * @param channel Channel to connect
      * @return Subscription instance
      */
-    fun create(channel: Channel): Subscription = Subscription(consumer, channel).also {
-        subscriptions.add(it)
+    fun create(channel: Channel): Subscription {
+        require(channel.identifier !in subscriptions.map { it.identifier }) { "Such subscription already exists" }
```
Comment (neworld) [inline library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt]: I see the `identifier` contains the name of the channel and other parameters. Different order of parameters leads to a different identifier. Also, created a new `Channel` with the same parameters in the same order leads to a crash. I believe you could get many confusions here.

Comment (neworld) [inline library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt]: Your implementation requires to create a new array due to mapping. You could improve performance without sacrificing readability

```suggestion
        require(subscriptions.none { it.identifier == channel.identifier) { "Such subscription already exists" }
```

Comment (neworld) [inline library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt]: Typically libraries are intended to be thread-safe. Your ir not. Two threads could create a subscription for the same channel.

File: library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt [hunk 2]
```diff
@@ -33,7 +40,20 @@ class Subscriptions constructor(private val consumer: Consumer) {
...
+            subscription.onDisconnected?.invoke()
+        }
+    }
+
+    /**
+     * Remove all subscriptions.
+     */
+    fun removeAll() {
+        subscriptions.forEach { subscription ->
+            consumer.send(Command.unsubscribe(subscription.identifier))
```
Comment (neworld) [inline library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt]: You should DRY implementation detail and call `remote(subscription)`

File: library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt [hunk 3]
```diff
@@ -2,52 +2,135 @@ package com.vinted.actioncable.client.kotlin
...
-import kotlinx.coroutines.runBlocking
-import kotlinx.coroutines.withTimeout
 import okhttp3.mockwebserver.MockWebServer
-import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.*
 import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
 import java.net.URI
 
 class SubscriptionsTest {
```
Comment (neworld) [inline library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt:12]: Please write a two tests with:

```
    private val channel1 = Channel(channel = "foo", params = mapOf("param1": "true"))
    private val channel2 = Channel(channel = "foo", params = mapOf("param1": "false"))
```

and with:

```
    private val channel1 = Channel(channel = "foo", params = mapOf("param1": "true", "param2": "false"))
    private val channel2 = Channel(channel = "foo", params = mapOf("param2": "false", "param1": "true"))
```

File: library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt [hunk 4]
```diff
@@ -2,52 +2,135 @@ package com.vinted.actioncable.client.kotlin
...
+                }
+            }
+            consumer.subscriptions.removeAll()
+
+            assertFalse(consumer.subscriptions.contains(subscription1))
+            assertFalse(consumer.subscriptions.contains(subscription2))
+            assertEquals(SUCCESSFUL_DISCONNECTION_MESSAGE, events.receive())
+            assertEquals(SUCCESSFUL_DISCONNECTION_MESSAGE, events.receive())
+            events.close()
+            Unit
```
Comment (GretaGri) [inline library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt]: > Unit 

Why this line needed?

**Comment by GretaGri** [comment]
After those updates README could be updated too, describing the current behaviour, additional functionality.

File: library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt (hunk 4 above)
Comment (GediminasZukas) [reply library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt]: Removed

File: library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt (hunk 3 above)
Comment (GediminasZukas) [reply library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt:12]: What kind of tests? To verify that such channels are not identical?

Comment (neworld) [reply library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt:12]: To know what is the expected behavior in this ambiguity.

File: library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt (hunk 1 above)
Comment (GediminasZukas) [reply library/src/main/kotlin/com/vinted/actioncable/client/kotlin/Subscriptions.kt]: Removed this functionality. User can have multiple subscriptions to the same channel if needed. Added a check ```contains(channel: Channel)``` to check if at least single subscription exists for the channel.

File: library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt (hunk 3 above)
Comment (GediminasZukas) [reply library/src/test/kotlin/com/vinted/actioncable/client/kotlin/SubscriptionsTest.kt:12]: Added some multiple channels identity tests.

**Review by neworld — APPROVED** [review]

**Comment by GediminasZukas** [comment]
> Why do you need toMutableList()? forEach is implemented for all iterators and collections

I need a copy of original collection for items removal in original collection during iteration. Otherwise, removal won't work.

Maybe I can change to ```ArrayList(subscriptions).forEach(::remove)``` for more convenience

**Pull request merged**